src/rust/bitbox02-noise/src/x25519.rs

```diff
@@ -0,0 +1,61 @@
+// Copyright 2020 Shift Cryptosecurity AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! This module implements the X25519 trait needed by noise_protocol
+//! by using the x25519_dalek crate. It is adapted from
+//! https://github.com/sopium/noise-rust/blob/76fb694f06b429879c264087f496958a99710356/noise-rust-crypto/src/lib.rs#L31,
+//! but uses a pluggable random number generator to generate keys.
+
+use core::ops::{Deref, DerefMut};
+use noise_protocol::U8Array;
+use noise_rust_crypto::sensitive::Sensitive;
+
+pub trait Random32 {
+    fn mcu_32_bytes(out: &mut [u8; 32]);
+}
+
+pub struct X25519<R: Random32>(core::marker::PhantomData<R>);
+
+pub type PrivateKey = [u8; 32];
+pub type PublicKey = [u8; 32];
+
+impl<R: Random32> noise_protocol::DH for X25519<R> {
+    type Key = Sensitive<PrivateKey>;
+    type Pubkey = PublicKey;
+    type Output = [u8; 32];
+
+    fn name() -> &'static str {
+        "25519"
+    }
+
+    fn genkey() -> Self::Key {
+        let mut k = Self::Key::new();
+        R::mcu_32_bytes(k.deref_mut());
+        k[0] &= 248;
+        k[31] &= 127;
+        k[31] |= 64;
```

Comment:
The top of the file has a comment that links to the file where the whole thing was taken from. Not enough?

Reply:
That link may become dead and their code doesn't explain it for us, mere mortals, either anyway. I'd prefer to be explicit - it doesn't cost much, does it.

Reply:
> Would be slightly easier to read if these were in binary like `0b11111000`

Won't change that as it is used like this also in x25519-dalek.

Added this comment now:

```
// Copied from: https://github.com/sopium/noise-rust/blob/76fb694f06b429879c264087f496958a99710356/noise-rust-crypto/src/lib.rs#L49-L51
        // which in turn copied it from:
        // https://github.com/dalek-cryptography/x25519-dalek/blob/ecd6be674850a99ad26404f6aa29b0cf79642b97/src/x25519.rs#L162-L164
        // which is also in our vendored deps: `vendor/x25519-dalek/src/x25519.rs`.
```

even if the links become dead, the pointer to the vendored file will remain.

Reply:
When I copy code it becomes something I own and have to maintain regardless of the original, so I'd want to understand exactly what it does, esp. after 6 month or a year. "Won't change that as it is used like this also in x25519-dalek." wouldn't really justify it for me. Do you always keep code exactly as is when copied from stackoverflow? :)

I guess the pointer to the vendored version will do. Thanks!

Reply:
> wouldn't really justify it for me. Do you always keep code exactly as is when copied from stackoverflow? :)

> I'd want to understand exactly what it does, esp. after 6 month or a year. 

If you change the code but the link to the source is still there, it is extra confusing and takes more time to re-evaluate it.

Besides, there is really no reason to change a cryptographic key generation function from a crypto source. There is no benefit, but a non zero risk of introducing an error.

Reply:
Oh sure, I'm fine with leaving it in decimal. I was mostly complaining about:

> The top of the file has a comment that links to the file where the whole thing was taken from. Not enough?

because imho it isn't enough: comments like "copied from here and there" with links which don't explain what it's for and why make it harder to understand and maintain in the future.